engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Type.java

```diff
@@ -34,6 +34,7 @@ public final class Type implements EnsoObject {
   private final Type supertype;
   private final Type eigentype;
   private final Map<String, AtomConstructor> constructors;
+  private boolean hasProjectPrivateConstructors;
```

Comment:
Could this field be `final` and be specified when the `Type` is being created?

Reply:
How are extension methods related to knowing constructors?

I don't think we can add an 'extension-constructor' thru the extension mechanism, can we?

Reply:
If we move this field to a `Type` (and rename to `hasPrivateConstructors`), then it shall be possible to make it `final`. Inside of `IrToTruffle` one has access to all constructors of a `type` when registering the type and check check if they are `private` or not.